### ISSUE TITLE
Please remove the ActiveSupport warning

### ISSUE THREAD

**Issue by DannyBen** [sinatra/sinatra#1476]
https://github.com/sinatra/sinatra/blob/b7c10649cad56f3263df089027621fb1753620cb/lib/sinatra/indifferent_hash.rb#L2-L8

Sinatra is used a lot as an embedded server. Now some of the tools that I have developed display this warning in the end result (i.e. to the user, not the developer).

I am not even including ActiveSupport directly in my gem, it is included through sinatra-contrib....

This type of message is suitable for documentation, not for runtime in my opinion.

If you disagree, at least please give a developer a way to disable it, please...

**Comment by mwpastore** [comment]
> If you disagree, at least please give a developer a way to disable it, please...

Would an environment variable be an acceptable way of doing this?

**Comment by DannyBen** [comment]
If you ask me, this "fix" creates more issues than it solves, and users should not be required to do *anything* to make it stop showing a warning, that they can do nothing about... 

Don't get me wrong, I have mad respect for all the free time people put into open source, and Sinatra is one of my favorite Ruby projects, but - as an open source developer myself - I think we all need to remember that people are depending on what we do, and printing developer related warnings to stdout seems a little aggressive and out of place.

That said, if I failed to "pull you to my side of the table" - yes, an environment variable was what I had in mind, since it can be easily set up at the top level of anything that uses it, without touching any intermediate code layers.

**Comment by mwpastore** [comment]
There are some horrifyingly subtle bugs that can happen here if you don't get the code ordering _just so_. It didn't make sense (to me) to bury the warning in a piece of documentation that no one will ever read. I agree that it is aggressive and ugly, but I disagree that it "creates more issues than it solves." 

I'm definitely open to alternative solutions and will defer to the Sinatra team on this issue. An opt-out environment variable is probably the easiest and most portable "workaround," if a bit kludgy. We could only show the warning in dev and/or test. We could print to stderr instead of stdout. Or some combination of these.

A neat solution would be to defer the check until the app is completed booted and _then_ compare the signatures of Hash and Sinatra::IndifferentHash to see if the former has been monkey-patched but not the latter (e.g. `#slice` is defined on Hash but not Sinatra::IndifferentHash), indicating a condition where the aforementioned footguns may be possible. However, I don't believe there's an easy, portable way to hook onto an "app is finished loading" event.

**Comment by DannyBen** [comment]
Ok..... just note that for me at least, changing the warning to only show in non production is not going to cut it. In my command line utilities that happen to have Sinatra as the driver behind one of their sub commands, this warning will always show - like this:

```shell
$ madness --help
WARNING: If you plan to load any of ActiveSupport's core extensions to Hash, be
sure to do so *before* loading Sinatra::Application or Sinatra::Base. If not,
you may disregard this warning.
Madness

Usage:
  madness [PATH] [options]
  madness create config
  madness create theme FOLDER
  madness (-h|--help|--version)

...(truncated)...
```

**Comment by mwpastore** [comment]
We only need to show the warning message on older Rubies, so maybe that's a step we can take immediately. What version of Ruby are you running under?

**Comment by DannyBen** [comment]
> We only need to show the warning message on older Rubies

This sounds great. Which is the highest version that requires this warning?
I am personally using 2.5.1, but I develop gems that support back to the area of 2.2.2 usually.

**Comment by mwpastore** [comment]
Anything under 2.5 could potentially be monkeypatched by ActiveSupport and break IndifferentHash.

**Comment by sammyhenningsson** [comment]
I don't understand the reason for this. My tests are now failing in TravisCI since I don' have activesupport installed. If activesupport is now a required dependency, then please add it that to the gemspec!!

**Comment by mwpastore** [comment]
The reason for what? ActiveSupport is not a required dependency for
Sinatra. It *is* a required dependency for Sinatra::Contrib, and it's in
the gemspec for that gem.

**Comment by sammyhenningsson** [comment]
Now that ActiveSupport is used in IndifferentHash I think it should be a dev dependency for Sinatra.
I have a project that does not use neither Sinatra::Contrib nor ActiveSupport. Since 2.0.4, my tests are now failing because Sinatra is trying to load ActiveSupport.

**Comment by mwpastore** [comment]
@sammyhenningsson ActiveSupport is **not** used by IndifferentHash and it was never intended to be a dependency, but yes, there is a bug in the logic that conditionally loads ActiveSupport, and it should be resolved by #1477.

**Comment by npickens** [comment]
Alternative solution: **get rid of the ActiveSupport dependency entirely.**

* The use of `IndifferentHash` in `sinatra-contrib/config_file` would be trivial to handle in-line.
* `sinatra-contrib/capture`'s use of `try` would also be easy to replace with simple `respond_to` checks ([related comment](https://github.com/sinatra/sinatra/commit/4b77084a178665c3732ee7251d15cf18ed2fea6c#commitcomment-29907022)).

Am I missing other uses?

**Comment by ginjo** [comment]
What if the warning were to show ONLY when the gems load in a way that could cause problems? And if we follow the warning's instructions to load the gems in a specific order, the warning would go away.

Would this even be possible, or would it create as many problems as it solves?

**Comment by mwpastore** [comment]
> What if the warning were to show ONLY when the gems load in a way that could cause problems?

This is impossible.

> And if we follow the warning's instructions to load the gems in a specific order, the warning would go away.

If you choose to load ActiveSupport core extensions to hash, and load them before requiring Sinatra, the warning will not show.

**Comment by SwagDevOps** [comment]
Please remove ActiveSupport.

**Comment by ZimbiX** [comment]
Here's the least uncomfortable workaround I could think of:

```ruby
# Suppress Sinatra's Active Support warning by tricking it into thinking Active Support's Hash extensions are already loaded
# TODO: Remove once the Sintra resolves the issue: https://github.com/sinatra/sinatra/issues/1476
$LOADED_FEATURES << 'fake/active_support/core_ext/hash'
``` 

Once Active Support is removed from sinatra-contrib, this will be irrelevant.

**Comment by namusyaka** [comment]
Closed via #1477
